File: app/src/main/java/com/example/vladislav/androidstudy/services/ServiceDemo2.java
package com.example.vladislav.androidstudy.services;

import android.app.Service;
import android.content.Intent;
import android.os.IBinder;
import android.support.annotation.Nullable;
import android.util.Log;
import android.widget.Toast;

/**
 * Created by vladislav on 18.05.17.
 * Description is at @strings/service_demo2_description
 */

public class ServiceDemo2 extends Service {

    final String LOG_TAG = "ServiceDemo2";

    public void onCreate() {
        super.onCreate();
        Log.i(LOG_TAG, "onCreate");
    }

    public int onStartCommand(Intent intent, int flags, int startId) {
        Log.i(LOG_TAG, "onStartCommand");
        someTask();
        return super.onStartCommand(intent, flags, startId);
    }

    public void onDestroy() {
        super.onDestroy();
        Log.i(LOG_TAG, "onDestroy");
    }

    @Override
    @Nullable
    public IBinder onBind(Intent intent) {
        Log.i(LOG_TAG, "onBind");
        return null;
    }

    void someTask() {
        try {
            Toast.makeText(this, "Service demo2 performs", Toast.LENGTH_SHORT).show();
            Thread.sleep(5000);
            Toast.makeText(this, "Service demo2 is over", Toast.LENGTH_SHORT).show();
        } catch (InterruptedException e) {
            e.printStackTrace();
        }
        stopSelf();
    }
}

File: android_study/src/main/java/com/example/vladislav/androidstudy/services/ServiceDemo2.kt
package com.example.vladislav.androidstudy.services

import android.app.Service
import android.content.Intent
import android.os.IBinder
import android.util.Log
import android.widget.Toast
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext

/**
 * Simple service demo to show its callbacks.
 */
class ServiceDemo2 : Service() {

    private val job = SupervisorJob()
    private val scope = CoroutineScope(Dispatchers.Main + job)

    override fun onCreate() {
        super.onCreate()
        Log.i(LOG_TAG, "onCreate")
    }

    override fun onStartCommand(intent: Intent?, flags: Int, startId: Int): Int {
        Log.i(LOG_TAG, "onStartCommand")
        someTask()
        return super.onStartCommand(intent, flags, startId)
    }

    override fun onDestroy() {
        super.onDestroy()
        // Cancelling job for coroutine could not last longer than life span of this Service.
        job.cancel()
        Log.i(LOG_TAG, "onDestroy")
    }

    override fun onBind(intent: Intent?): IBinder? {
        Log.i(LOG_TAG, "onBind")
        return null
    }

    // UI freezes
    private fun someTask() {
        Toast.makeText(this@ServiceDemo2, "ServiceDemo2 is started", Toast.LENGTH_SHORT).show()
        Thread.sleep(3000)
        Toast.makeText(this@ServiceDemo2, "ServiceDemo2 is over", Toast.LENGTH_SHORT).show()
    }

    // Just to be aware of how to make UI not freezing
    private fun someTask2() {
        // This scope runs on a main thread, since one should show Toast,
        scope.launch {
            Toast.makeText(this@ServiceDemo2, "ServiceDemo2 is started", Toast.LENGTH_SHORT).show()
            // but delay should go on a separate thread, for UI could not be locked.
            withContext(Dispatchers.IO) {   // Dispatchers should be provided through DI
                delay(3000)
            }
            Toast.makeText(this@ServiceDemo2, "ServiceDemo2 is over", Toast.LENGTH_SHORT).show()
        }
    }

    companion object {
        private const val LOG_TAG = "ServiceDemo2"
    }
}